Build: Support referencing GirCore via project reference

badcel opened the pull request:
To enable easy debugging of GirCore libraries, the msbuild property "GirCoreSource" can be set to a path where the GirCore source is located.

If this property is set the GirCore package references will be replaced by project references which point to the local GirCore source.

badcel commented:
@cameronwhite: This would be the first solution that I could figure out to easily switch between package and project references. The code is not complete and only done for libadwaita. Additionally I purposely left the path to my local GirCore source inside `Directory.Build.props` to show you how it could look like if configured for local source use.

This should result in project references be used (which breaks the build as no source is available under the given path). If you remove the path from the tag the package reference should be used instead, including CentralPackageManagement.

This approach creates one target file for each GirCore package. Each project which wants to have a certain GirCore package must import this target.

An alternative would be to define the code of the targets file in each csproj file.

I don't know what I like better but I think this shows the idea best as it deduplicates code but adds a little complexity to the build as the `ProjectReference` and `PackageReference` tags are hidden inside custom targets.

Additionally I added `Directory.Build.props`, `Directory.Packages.props`, and the new target  file to the `Solution Items` solution folder which is just my preference to be able to easily reach those files without accessing the file system if Rider is in the "SolutionView". I can of course remove it if you prefer.

cameronwhite commented:
Thanks for looking into this!
I'm pretty happy with the approach - I like that it doesn't add much complexity to the individual .csproj files and the package version is still managed in `Directory.Packages.Props` like the other packages

No problem with adding extra items to the `Solution Items` folder 👍

badcel commented:
I wonder if the name of the targets file should be `AddReferenceGirCoreAdw1.target` ("target" instead of "targets" and a little more expressive)?

And the target inside would match the filename?

cameronwhite commented:
Yeah the `target` filename sounds better to me :+1: 
Does the target name show up anywhere in the UI for IDEs? If so then matching the filename sounds reasonable

badcel commented:
I just saw that you only reference 4 packages from GirCore: adwaita, Gtk, GObject-2.0.Integration, PaingoCairo-1.0

But you are using some implicit dependencies which get pulled into the project from the nugets like: Gdk and GLib. For this to work with project references those would need to be added explicitly. Is this OK for you?

cameronwhite commented:
I think that's okay to make the dependencies more explicit 👍

badcel commented:
The usage of targets made more problems than it helped. The files now got even simpler and just define package / project references as project properties. So the files got renamed to reflect this change.

What do you think? 

Remark: In Rider I could not fluently switch from the nuget packages to the local repository. I had to do a manual compile on the commandline first (`dotnet build`). I don't know if this is perhaps a bug in Rider? Afterwards it works in Rider, too.

cameronwhite commented:
Thanks, I'm good with the updated approach! Agreed that seems a bit simpler. 👍 
We'd still need to add all the implicit dependencies, correct?

badcel commented:
No it looks like it is not needed. When I added all implicit dependencies the errors went away in rider. But compiling on the command line failed nevertheless.

Afterwards I removed all the target stuff and it worked without defining all dependencies.

So I removed it again as it was quite a bit of files and a lot of changes.

From my point of view you could test if this PR works for you.

cameronwhite commented:
Thanks! I've tested building against my local copy of gir.core and it worked well for me, so I think this should be good to merge 🎉
